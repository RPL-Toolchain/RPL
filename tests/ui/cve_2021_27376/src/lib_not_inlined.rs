//@compile-flags: -Z inline-mir=false
extern crate libc;

use libc::{sockaddr, sockaddr_storage, socklen_t};
use std::mem::{self, MaybeUninit};
use std::net::{SocketAddr, SocketAddrV4, SocketAddrV6};
use std::ptr;

struct Addr {
    storage: sockaddr_storage,
    len: socklen_t,
}

impl Addr {
    /// Creates a raw socket address from `SocketAddr`.
    // #[rpl::dump_mir(dump_cfg, dump_ddg)]
    fn new(addr: std::net::SocketAddr) -> Self {
        let (addr, len): (*const sockaddr, socklen_t) = match &addr {
            SocketAddr::V4(addr) => (
                addr as *const SocketAddrV4 as *const sockaddr,
                //~^ ERROR: wrong assumption of layout compatibility from `std::net::SocketAddrV4` to `libc::sockaddr`
                mem::size_of_val(addr) as socklen_t,
            ),
            SocketAddr::V6(addr) => (
                addr as *const SocketAddrV6 as *const sockaddr,
                //~^ ERROR: wrong assumption of layout compatibility from `std::net::SocketAddrV6` to `libc::sockaddr`
                mem::size_of_val(addr) as socklen_t,
            ),
        };
        unsafe { Self::from_raw_parts(addr, len) }
    }

    /// Creates an `Addr` from its raw parts.
    unsafe fn from_raw_parts(addr: *const sockaddr, len: socklen_t) -> Self {
        let mut storage = MaybeUninit::<sockaddr_storage>::uninit();
        unsafe {
            ptr::copy_nonoverlapping(
                addr as *const u8,
                &mut storage as *mut MaybeUninit<sockaddr_storage> as *mut u8,
                len as usize,
            );
        }
        Self {
            storage: unsafe { storage.assume_init() },
            len,
        }
    }
}

fn main() {
    todo!()
}
