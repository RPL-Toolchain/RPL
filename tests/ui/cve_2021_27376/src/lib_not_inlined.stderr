error: wrong assumption of layout compatibility from `std::net::SocketAddrV6` to `libc::sockaddr`
  --> tests/ui/cve_2021_27376/src/lib_not_inlined.rs:25:17
   |
LL |                 addr as *const SocketAddrV6 as *const sockaddr,
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: it's not guaranteed by Rust standard library. See https://github.com/rust-lang/rust/pull/78802
note: casted from this
  --> tests/ui/cve_2021_27376/src/lib_not_inlined.rs:25:17
   |
LL |                 addr as *const SocketAddrV6 as *const sockaddr,
   |                 ^^^^

error: wrong assumption of layout compatibility from `std::net::SocketAddrV4` to `libc::sockaddr`
  --> tests/ui/cve_2021_27376/src/lib_not_inlined.rs:20:17
   |
LL |                 addr as *const SocketAddrV4 as *const sockaddr,
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: it's not guaranteed by Rust standard library. See https://github.com/rust-lang/rust/pull/78802
note: casted from this
  --> tests/ui/cve_2021_27376/src/lib_not_inlined.rs:20:17
   |
LL |                 addr as *const SocketAddrV4 as *const sockaddr,
   |                 ^^^^

error: aborting due to 2 previous errors

